### PR TITLE
Add a nix flake providing a development shell

### PR DESCRIPTION
**Pull request by charludo**
### Short description

(Since this has nothing to do with any issue or any plans we've discussed as a team, I feel that I need to preface this with: everything related to this was done on my own time, not during working hours.)

This PR adds a [nix flake](https://wiki.nixos.org/wiki/Flakes). It allows developers to utilize a standardized and completely reproducible development environment.

You might find this useful if you currently manage multiple python and/or node projects (possibly requiring conflicting versions), are a nix user already, or frequently re-setup the project; but more importantly I strongly believe this makes setting our project up significantly easier for new developers (speaking from experience from my own onboarding two years ago 😉):

Regardless of the system a person is on - any kind of Linux; MacOS; even Windows with WSL enabled - getting up and running with our project would boil down to

1.  clone this repo
2. install nix (and **only** nix, not python/node/...), for example with the awesome installer script from [zero-to-nix.com](https://zero-to-nix.com):
    ```bash
    curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix | sh -s -- install
    ```
3. run `nix develop` in your shell

As a result, you'll have the exact same development environment available as me/anyone else following these steps.

### Proposed changes

- add a nix flake and corresponding lock file 

### Side effects

- none, except:
- I volunteer to give a bar camp about nix/nix develop/NixOS at the next conference 😄 

### Resolved issues

Fixes: n/a

__________________________________________________

[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            python311Packages.magic
+            python311Packages.virtualenv
+
+            nodePackages.npm
+            nodejs_21
+
+            gettext
+            netcat-gnu
+            pcre16
+            file
```
Comment (deen13) [inline flake.nix:26]: Hey @charludo, this looks like a very beneficial addition. As someone currently undergoing the onboarding process, I recently set up my project and encountered a few hurdles, two of which would be addressed by this pull request. Let me outline the issues I faced, which I believe might also affect future macOS users:

- I initially encountered compatibility issues due to a newer Python version, which took some time to resolve. Fortunately, this pull request resolves this issue by pinning the Python version.
- On macOS, the default sed version differs from the one expected by translate.th, leading to potential issues. Remedying this typically involves manually installing GNU sed using brew install gnu-sed, or alternatively, incorporating it into this script as suggested. This [asciinema](https://asciinema.org/a/0mDY357AXvo5s55C7vhBauqEo) demonstrates that the issue is successfully resolved upon executing nix develop.
- Additionally, my macOS setup lacked the libmagic library, which is utilized by certain components of the system. Unfortunately, there isn't a Nix package readily available to provide this library, so this particular issue remains unresolved for now.

```suggestion
            file
            gnused
```

In summary integrating gnu-sed into the script would significantly alleviate macOS-related compatibility issues encountered by users. Moreover, if we were to develop and publish a Nix package for libmagic and incorporate it into this pull request, we could effectively resolve all remaining macOS-related hurdles. 😌

Comment (charludo) [reply flake.nix:26]: Hi @deen13, thanks for the feedback, very glad to hear this is already useful!! 😄

Funnily enough, I did also encounter libmagic-related issues when attempting to get the project running under NixOS, but this flake did fix those issues for me.

I will have another look at that specifically. Did you resolve your issues by just installing a MacOS version of libmagic globally? Or was there another workaround?

Comment (charludo) [reply flake.nix:26]: @deen13 unfortunately I don't have MacOS available to test this.

Would you be OK checking if adding
```nix
python311Packages.magic
python311Packages.python-magic
python311Packages.typecode-libmagic
python311Packages.typecode
```

To the packages list resolves the issue? 🙈

Comment (deen13) [reply flake.nix:26]: Hey @charludo, sure I'll check if it works once I'm at my computer later this day. I resolved this issues, as you already said, by installing libmagic globally via brew.

Comment (deen13) [reply flake.nix:26]: Hey @charludo, [nix develop](url) is failing on my system with an error, which you can see in the attached screenshot. It appears that it's attempting to use a native version of libmagic. What are your thoughts?

image

Comment (charludo) [reply flake.nix:26]: Hm, looks like `python311Packages.typecode-libmagic` is just [not packaged for MacOS](https://github.com/NixOS/nixpkgs/blob/1487bdea619e4a7a53a4590c475deabb5a9d1bfb/pkgs/development/python-modules/typecode/libmagic.nix#L48) :(

I don't think we should attempt allowing unsupported systems's inputs, so unfortunately my only idea is to check if omitting that package does the job still, and same thing if the build fails again due to `python311Packages.typecode` (which it might, apparently it pulls in `python311Packages.typecode-libmagic` anyways).

Maybe `python311Packages.python-magic` is sufficient to solve the problem? 

Sorry for being so vague, without a system to test this on I'm tapping in the dark :(

Comment (deen13) [reply flake.nix:26]: I've already tried `python311Packages.python-magic` and this does not work neither. I totally understand the problems you're facing w/o having a mac to test on. I'm currently playing with devcontainers in order to solve the issues specifically for mac users. Maybe in the future nix flakes and devcontainers can co-exist happily. 🤗

Comment (deen13) [reply flake.nix:26]: Yesss that's all. ☺️

Comment (charludo) [reply flake.nix:59]: Sorry to bother you again @deen13, but would you mind trying the flake again? I was able to replicate the missing `libmagic` by entering the shell in "pure" mode (`nix-develop --unset PATH`), and have been able to fix the issue (by passing `file` into the `LD_LIBRARY_PATH`). 

This + the new automatic postgres setup *should* solve the remaining issues.

Comment (deen13) [reply flake.nix:59]: Hey @charludo, I pulled the latest changes and ran `nix develop && ./tools/run.sh` again but it still fails to find libmagic. :/

File: README.md [hunk 2]
```diff
@@ -36,6 +36,34 @@ cd integreat-cms
...
 ### Run development server
 
+### Choosing your setup method
+
+From the three provided development setup options, choose yours based on personal preference and familiarity with the used tools. If you are unsure, the following might help you make a decision.
+Please note that these are only suggestions.
+- using VSCode or PyCharm? → [Devcontainer](#user-content-devcontainer)
+- on NixOS or already using `nix`? → [Nix Flake](#user-content-nix-flake)
+- conflicting version of tools required on your system? → [Devcontainer](#user-content-devcontainer) or [Nix Flake](#user-content-nix-flake)
+- do not want to use additional tooling and have no conflicting tooling? → [Local Setup](#user-content-local-setup)
```
Comment (deen13) [inline README.md:46]: Love that!

File: tools/prune_database.sh [hunk 3]
```diff
@@ -8,16 +8,22 @@
 source "$(dirname "${BASH_SOURCE[0]}")/_functions.sh"
 
 # Check if local postgres server is running
-if nc -w1 localhost "${INTEGREAT_CMS_DB_PORT}"; then
+if nc -z -w1 localhost "${INTEGREAT_CMS_DB_PORT}"; then
```
Comment (charludo) [inline tools/prune_database.sh:11]: With docker, the `z` flag did not make a difference in my testting, but ith psql, the absence of `-z` led to `nc` never exiting.

**Review by deen13 — COMMENTED** [review]
Looks good code-wise but I ran into an error when starting the development server: 

```
✔ The Integreat CMS was successfully installed 😻
Use the following command to start the development server:

	./tools/run.sh

(nix:nix-shell-env) (.venv) mbp:integreat-cms deen$ ./tools/run.sh 
Checking if Integreat CMS is installed...
✔ Integreat CMS is installed
Checking if local Redis server is running...
❌No Redis server detected. Falling back to local-memory cache.
pg_ctl: directory "/Users/deen/Coding/integreat-cms/.postgres/data" does not exist
(nix:nix-shell-env) (.venv) mbp:integreat-cms deen$ 
```

**Review by deen13 — APPROVED** [review]
Works like a charm 😌

**Comment by charludo** [comment]
@svenseeberg @timobrembeck @david-venhoff the three of you seemed most interested at the conference, would one of you mind doing the second review? 😅

**Comment by david-venhoff** [comment]
I also get an error
image

The env seems to contain these values though :thinking: 
image

**Comment by charludo** [comment]
> I also get an error image
> 
> The env seems to contain these values though 🤔 image

Interesting. Not sure why that would happen, but I'll have a look and try to find the source of the error.

Did `./tools/install.sh` work without errors?

**Comment by charludo** [comment]
@david-venhoff [this SO post](https://stackoverflow.com/questions/50746147/postgresqls-initdb-fails-with-invalid-locale-settings-check-lang-and-lc-e) seemed relevant, describing the exact same issue with postgres. The first answer is some hacky workarounds....

...and the second answer is "you didn't mention nix, but I'm guessing this is a nix issue" 😂 

I've added `glibcLocales` to the packages list, would appreciate if you could check if the error is now resolved.

**Comment by david-venhoff** [comment]
Yes, the commit fixes my error. Now I get another one:

```
(.venv) user@laptop-pop-os:~/dev/integreat-cms$ ./tools/run.sh 
Checking if Integreat CMS is installed...
✔ Integreat CMS is installed
Checking if local Redis server is running...
✔ Running Redis server on port 6379 detected. Caching enabled.
✔ Running PostgreSQL database detected
Migrating database...
Traceback (most recent call last):
  File "/home/user/dev/integreat-cms/.venv/bin/integreat-cms-cli", line 54, in <module>
    main()
  File "/home/user/dev/integreat-cms/.venv/bin/integreat-cms-cli", line 50, in main
    execute_from_command_line(sys.argv)
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 416, in execute
    django.setup()
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/django/apps/registry.py", line 91, in populate
    app_config = AppConfig.create(entry)
                 ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/django/apps/config.py", line 123, in create
    mod = import_module(mod_path)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.pyenv/versions/3.11.9/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/user/dev/integreat-cms/integreat_cms/google_translate_api/apps.py", line 15, in <module>
    from google.cloud import translate_v2, translate_v3  # type: ignore[attr-defined]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/google/cloud/translate_v2/__init__.py", line 18, in <module>
    from google.cloud.translate import __version__
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/google/cloud/translate/__init__.py", line 21, in <module>
    from google.cloud.translate_v3.services.translation_service.async_client import (
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/google/cloud/translate_v3/__init__.py", line 21, in <module>
    from .services.translation_service import (
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/google/cloud/translate_v3/services/translation_service/__init__.py", line 16, in <module>
    from .async_client import TranslationServiceAsyncClient
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/google/cloud/translate_v3/services/translation_service/async_client.py", line 32, in <module>
    from google.api_core import gapic_v1
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/google/api_core/gapic_v1/__init__.py", line 16, in <module>
    from google.api_core.gapic_v1 import config
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/google/api_core/gapic_v1/config.py", line 23, in <module>
    import grpc
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/grpc/__init__.py", line 22, in <module>
    from grpc import _compression
  File "/home/user/dev/integreat-cms/.venv/lib/python3.11/site-packages/grpc/_compression.py", line 20, in <module>
    from grpc._cython import cygrpc
ImportError: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.38' not found (required by /nix/store/c2yb135iv4maadia5f760b3xhbh6jh61-gcc-13.2.0-lib/lib/libstdc++.so.6)
```

I found [this](https://discourse.nixos.org/t/new-install-nix-wants-glibc-2-38-debian-12-has-2-36/36109/4) post which suggests to `unset LD_LIBRARY_PATH` and then the server finally starts on my laptop. I don't really know what this means though or where that environment variable gets set :shrug:

**Comment by charludo** [comment]
> I found [this](https://discourse.nixos.org/t/new-install-nix-wants-glibc-2-38-debian-12-has-2-36/36109/4) post which suggests to `unset LD_LIBRARY_PATH` and then the server finally starts on my laptop. I don't really know what this means though or where that environment variable gets set 🤷

We actually set that ourselves, currently:
```bash
LD_LIBRARY_PATH = pkgs.lib.makeLibraryPath [ pkgs.stdenv.cc.cc.lib pkgs.file ];
```

This is necessary to make `libstdc++` available to the python version which gets linked in the process of creating a `venv`. As far as I can tell, the "setting `LD_LIBRARY_PATH` at *all* is a massive red flag" comment from the thread you linked refers to the context of the original post, where `LD_LIBRARY_PATH` is set on a system-level, which is indeed heavily discouraged.

Inside of development shells on the other hand, it's the preferred way to get python venvs working (see for example [this thread](https://discourse.nixos.org/t/how-to-solve-libstdc-not-found-in-shell-nix/25458/7), where this distinction is also made).

However, I just realized that `stdenv.cc.cc.lib` is not part of the input packages. I've added it, hopefully this will make the required `glibc` available on non-nix systems (btw, not exporting `LD_LIBRARY_PATH` leads to very similar errors as you've posted above for me 😂 I'm guessing that the `glibc` existing on my system is always the required one (since coming from `nixpkgs` both for my system, and for the packages installed in the shell), while the one provided by your distribution differs from the one the packages in the shell rely on).

In any case, I'm sorry that I can't be more certain if this will fix it, since I can't reproduce the error 🙈

**Review by david-venhoff — APPROVED** [review]
Nice! :tada:
